compiler/rustc_monomorphize/src/partitioning.rs

```diff
@@ -254,8 +254,9 @@ where
             always_export_generics,
         );
 
-        // We can't differentiate something that got inlined.
+        // We can't differentiate a function that got inlined.
         let autodiff_active = cfg!(llvm_enzyme)
+            && matches!(mono_item, MonoItem::Fn(_))
```

Comment:
Discussion: this kinda feels backwards to me, I feel like if autodiff is applied to some fn it should inhibit inlining. Is there a small test case (synthetic or not) that can demonstrate this change?

Reply:
I add inline(never) in my frontend macro to source and target function. However, those aren't guaranteed to prevent inlinung, so I additionally check it here in rustc_monomorphize. Every working codegen test shows that my logic here and in the frontend works. Adding additional must-fail tests would require building rustc with changes to the frontend (to not emit noinline) and the middle-end (to remove this code). That seems like a lot of effort to test. But I feel like I probably misunderstand your question?

The change btw. is needed since we currently also inspect non fn items like Asm, which triggers a debug assertion, causing bootstrap to fail. But you already mentioned that this looks reasonable to you

Reply:
> But I feel like I probably misunderstand your question?

I moreso meant if there's a smaller example that would fail without this PR, not changing the front-end/middle-end bits.

Reply:
aah, yes now I get your question. I only mentioned it in the title, not the PR body, but you can find the panic stacktrace in the linked issue.

Reply:
https://github.com/rust-lang/rust/pull/140064#issuecomment-2817989150

@jieyouxu since you already looked at it, it's a one line change, fixes bootstrap and eliminates the ad overhead, do you mind to rubberstamp it? Then I can drop it from my final enable-aitodiff PR.

Reply:
Yeah, at least this makes it possible to build, and I can't come up with a smaller test...